### PR TITLE
Issues script revamp

### PR DESCRIPTION
**Pull request by lexicalunit**
Updates the script to work with the new milestone process.

- [x] issues: fixes https://github.com/bokeh/bokeh/issues/10213

## Check Only Mode
- [x] Relevant issues/PRs are those for the current specified milestone
- [x] All issues in the milestone should be closed, with a reso:completed
- [x] All PRs in the milestone should be closed/merged with status:accepted
- [x] Any PR with a type label (an "issue PR") also needs reso:completed
- [x] There should only ever be one reso or status label applied.
- [x] All issues sholud have a type
- [x] Allowed types: bug, feature, task
- [x] Nothing should have TRIAGE label
- [x] All issues/PRs should be closed

## Update Mode
- [x] Should do all the checks from check only mode first
- [x] Group by bugfix, feature, task
- [x] Include component label, if present

### CODE REVIEW

File: scripts/milestone.py [hunk 1]
```diff
@@ -0,0 +1,433 @@
+#!/usr/bin/env python3
+
+import argparse
+import logging
+import sys
+
+import requests
+
+# from collections import OrderedDict
```
Comment (mattpap) [inline scripts/milestone.py:9]: Is this PR work in progress? Otherwise all the commented code should be removed.

**Comment by mattpap** [comment]
> Apply black formatting

We don't use black, so this step wasn't necessary.

**Comment by bryevdv** [comment]
> We don't use black, so this step wasn't necessary

FYI I added black style checking scoped only to the `release` sub-directory in #10239 but I would like to experiment with applying it to smaller more out of the way locations of the codebase (the `scripts` an `ci` sub-directories are obvious choices).

File: scripts/milestone.py (hunk 1 above)
Comment (lexicalunit) [reply scripts/milestone.py:9]: Yep I didn't have access to add the label myself. This will all be cleaned up eventually!

**Comment by lexicalunit** [comment]
@mattpap 

> We don't use black, so this step wasn't necessary.

Mostly just personal hygiene on my part. I don't write non-formated code anymore if I can help it. That commit was from when I thought I could mostly re-use the code and was my first step in getting reacquainted with it. Since moving to `milestone.py` I think much of the script will just have to be rewritten anyway so the point is moot.

**Comment by lexicalunit** [comment]
Ready for review.

File: scripts/milestone.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+
+    problems = check_milestone_items(items)
+    for problem in problems:
+        print(problem, file=sys.stderr)
+
+    if check_only:
+        if len(problems) > 0:
+            sys.exit(2)
+        else:
+            sys.exit(0)
```
Comment (bryevdv) [inline scripts/milestone.py:284]:
```suggestion
    if len(problems) > 0:
        sys.exit(2)
    elif check_only:
        sys.exit(0)
```

File: scripts/milestone.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+    "--check-only",
+    default=False,
+    is_flag=True,
+    help="Only verify the milestone for compliance, do not write to changelog",
+)
+def main(milestone, log_level, verbose, check_only):
+    """
+    Generates a bokeh changelog entry using the GitHub API.
+
+    Requires that you set GITHUB_API_TOKEN to your GitHub API Token. Exit code 2
```
Comment (bryevdv) [inline scripts/milestone.py:259]: Should be `GITHUB_TOKEN`

File: scripts/milestone.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
+#!/usr/bin/env python3
+
+import datetime
+import logging
+import os
+import sys
+from itertools import groupby
+
+import click
+import requests
```
Comment (bryevdv) [inline scripts/milestone.py:10]: I'm a little surprised `isort` is not compaining here about missing group headers, will have to investigate. Also will need to add `click` as dependency in conda environment yaml files

File: scripts/milestone.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+
+VALID_TYPES = [
+    "type: bug",
+    "type: feature",
+    "type: task",
+]
+
+
+def query_github(query, token):
+    """Hits the GitHub GraphQL API with the given query and returns the response."""
```
Comment (bryevdv) [inline scripts/milestone.py:20]: Out repo style is one space between triple quotes and the comment
```
""" Hits ...
```

File: scripts/milestone.py (hunk 4 above)
Comment (lexicalunit) [reply scripts/milestone.py:14]: Updated and added 👍

File: scripts/milestone.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+
+# External imports
+import click
+import requests
+
+VALID_TYPES = [
+    "type: bug",
+    "type: feature",
+    "type: task",
+]
```
Comment (bryevdv) [inline scripts/milestone.py:19]: nit: I prefer tuples for "constants" like this

Comment (lexicalunit) [reply scripts/milestone.py:19]: Yah that seems reasonable here 👍

File: scripts/milestone.py [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+            problems.append(f"issue PR has too many resolutions: {description(data)}")
+
+    # no prs with more than one status: label
+    if sum(1 for label in labels if label.startswith("status:")) > 1:
+        problems.append(f"PR has too many statuses: {description(data)}")
+
+    # no prs without a type: label
+    num_types = sum(1 for label in labels if label.startswith("type:"))
+    if num_types == 0:
+        problems.append(f"PR does not have a type: {description(data)}")
```
Comment (bryevdv) [inline scripts/milestone.py:184]: I think this block is wrong. PRs *may* have (one) type, those are "issue PRs". But normal PRs linked to a separate issue do not have a type.

Comment (lexicalunit) [reply scripts/milestone.py]: Fixed! I also had to update the code to account for PRs that don't have a type as you can't sort them with a key based on the value of their type, obviously. I also added a `-a` option to operate on closed milestones.

**Comment by bryevdv** [comment]
One macos job got stuck but it's not related to this code, will merge now. Thanks @lexicalunit !

**Pull request merged**